### ISSUE TITLE
Wrong P-Code for 67 E8 on 32-bit x86

### ISSUE THREAD

**Issue by nneonneo** [NationalSecurityAgency/ghidra#5666]
**Describe the bug**
In 32-bit Intel, the instruction `67 E8 xx xx xx xx` is incorrectly decoded as affecting `SS:SP` instead of `ESP`. This results in poor decompilation output.

This is very similar to #2976, but applies to 32-bit instructions.

**To Reproduce**
1. Open the attached `67call` binary in Ghidra. [67call.zip](https://github.com/NationalSecurityAgency/ghidra/files/12362400/67call.zip)
2. Observe the very convoluted decompilation, which includes `segment` calls, casting of `ESP` to `short`, and inability to properly recover function arguments:

```c
undefined4 main(undefined4 param_1)

{
  undefined4 *puVar1;
  undefined4 uVar2;
  int iVar3;
  short sVar4;
  undefined4 *puVar6;
  undefined2 in_SS;
  undefined4 *local_10;
  int *piVar5;
  
  local_10 = &param_1;
  sVar4 = (short)&local_10 + -4;
  piVar5 = (int *)CONCAT22((short)((uint)&local_10 >> 0x10),sVar4);
  puVar1 = (undefined4 *)segment(in_SS,sVar4);
  *puVar1 = 0x111c8;
  iVar3 = *piVar5;
  *piVar5 = param_1;
  piVar5[-1] = param_1;
  puVar6 = (undefined4 *)CONCAT22((short)((uint)(piVar5 + -1) >> 0x10),(short)(piVar5 + -1) + -4);
  puVar1 = (undefined4 *)segment(in_SS,(short)(piVar5 + -1) + -4);
  *puVar1 = 0x111da;
  uVar2 = foo();
  *puVar6 = uVar2;
  puVar6[-1] = iVar3 + 0xe40;
  puVar1 = (undefined4 *)segment(in_SS,(short)(puVar6 + -1) + -4);
  *puVar1 = 0x111ee;
  printf(*(char **)(CONCAT22((short)((uint)(puVar6 + -1) >> 0x10),(short)(puVar6 + -1) + -4) + 4));
  return 0;
}
```

The core cause is that the P-Code for the `67 E8` call instructions looks like this:

```
           000111d4 67 e8 c3        CALL          foo
                    ff ff ff
                                                      $U9680:4 = COPY 0x111da:4
                                                      SP = INT_SUB SP, 4:2
                                                      $U9800:4 = CALLOTHER "segment", SS, SP
                                                      STORE ram($U9800:4), $U9680:4
                                                      CALL *[ram]0x1119d:4
```

Running the binary on 32-bit Linux works as expected (it prints out `argc * argc`).

**Expected behavior**

Per the Intel SDM, the call instruction is only affected by the operand size attribute, not the address size attribute. Thus, `67 E8` should be decoded the same as `E8`, and should be using `ESP` instead of `SS:SP` in 32-bit mode.

When properly decoded, the decompilation should look like this:

```c
undefined4 main(undefined4 param_1)

{
  undefined4 uVar1;
  
  uVar1 = foo(param_1,param_1);
  printf("%d\n",uVar1);
  return 0;
}
```

and the P-Code should be

```
           000111d4 67 e8 c3        CALL          foo
                    ff ff ff
                                                      ESP = INT_SUB ESP, 4:4
                                                      STORE ram(ESP), 0x111da:4
                                                      CALL *[ram]0x1119d:4
```

**Environment (please complete the following information):**
 - OS: macOS 12.6.7
 - Java Version: 17.0.7
 - Ghidra Version: 10.3.2
 - Ghidra Origin: official GitHub distro

**Additional context**

IDA and Binary Ninja both handle these instructions correctly.

**Comment by GhidorahRex** [comment]
This is a known issue caused by the way we treat the `66` and `67` prefixes. Currently, the processor module doesn't have a way of knowing whether or not the instruction was modified with a prefix 67. Only what the current address size is. I think it's worth keeping this issue open so that we can track it however.

**Comment by nneonneo** [comment]
I think that's sensible - after all, 66/67 are explicitly defined as flipping the operand or address size parameters.

However, what's wrong here is that 67 should not affect CALL at all, because CALL is specified to only depend on OperandSize. This might be specific to CALL, so maybe it's enough to just modify the definition of CALL?

Here's the relevant part of the SDM:

```
IF near call
    THEN IF near relative call
        THEN
            IF OperandSize = 64
                THEN
                    tempDEST := SignExtend(DEST); (* DEST is rel32 *)
                    tempRIP := RIP + tempDEST;
                    IF stack not large enough for a 8-byte return address
                        THEN #SS(0); FI;
                    Push(RIP);
                    IF ShadowStackEnabled(CPL) AND DEST != 0
                        ShadowStackPush8B(RIP);
                    FI;
                    RIP := tempRIP;
            FI;
            IF OperandSize = 32
                THEN
                    tempEIP := EIP + DEST; (* DEST is rel32 *)
                    IF tempEIP is not within code segment limit THEN #GP(0); FI;
                    IF stack not large enough for a 4-byte return address
                        THEN #SS(0); FI;
                    Push(EIP);
                    IF ShadowStackEnabled(CPL) AND DEST != 0
                        ShadowStackPush4B(EIP);
                    FI;
                    EIP := tempEIP;
            FI;
            IF OperandSize = 16
                THEN
                    tempEIP := (EIP + DEST) AND 0000FFFFH; (* DEST is rel16 *)
                    IF tempEIP is not within code segment limit THEN #GP(0); FI;
                    IF stack not large enough for a 2-byte return address
                        THEN #SS(0); FI;
                    Push(IP);
                    IF ShadowStackEnabled(CPL) AND DEST != 0
                        (* IP is zero extended and pushed as a 32 bit value on shadow stack *)
                        ShadowStackPush4B(IP);
                    FI;
                    EIP := tempEIP;
            FI;
```

Compare with something like LEA, which has explicit checks for `OperandSize` and `AddressSize`:

```
IF OperandSize = 16 and AddressSize = 16
    THEN
        DEST := EffectiveAddress(SRC); (* 16-bit address *)
    ELSE IF OperandSize = 16 and AddressSize = 32
        THEN
            temp := EffectiveAddress(SRC); (* 32-bit address *)
            DEST := temp[0:15]; (* 16-bit address *)
        FI;
    ELSE IF OperandSize = 32 and AddressSize = 16
        THEN
            temp := EffectiveAddress(SRC); (* 16-bit address *)
            DEST := ZeroExtend(temp); (* 32-bit address *)
        FI;
...
```

**Comment by nneonneo** [comment]
FWIW, the stack address size (StackAddressSize/StackAddrSize in the x86 instruction pseudocode) is actually determined by the B bit of the SS segment descriptor in 16- or 32-bit mode - it is *not* determined by the AddressSize value. This is specified in the SDM volume 1, 6.2.3 "Address-Size Attributes for Stack Accesses". So, using `addresssize` to affect `push` or `pop` is incorrect in every case.

As it turns out, this issue also affects every other instruction that uses implicit push and pop, including CALLF, PUSH{,F,FD,FQ,A,AD}, POP{,F,FD,FQ,A,AD}, ENTER, LEAVE, IRET{,D,Q}, RET{,F}.

It's probably not unreasonable to just assume that the stack is as big as the operand size, although for maximum correctness we could have a separate "stack address size" bit in the context register which could be manually set by the analyst, and which defaults to the appropriate size for the instruction set.

**Comment by Wall-AF** [comment]
This affects #4685 also.

**Comment by GhidorahRex** [comment]
@nneonneo 
> I think that's sensible - after all, 66/67 are explicitly defined as flipping the operand or address size parameters.
>
>However, what's wrong here is that 67 should not affect CALL at all, because CALL is specified to only depend on OperandSize. This might be specific to CALL, so maybe it's enough to just modify the definition of CALL?

You're right that it shouldn't affect call, but the way our sleigh code is currently implemented, the CALL instruction has no method of determining whether an addrsize prefix has occurred or not. Only what the current address size is.

>FWIW, the stack address size (StackAddressSize/StackAddrSize in the x86 instruction pseudocode) is actually determined by the B bit of the SS segment descriptor in 16- or 32-bit mode - it is not determined by the AddressSize value. This is specified in the SDM volume 1, 6.2.3 "Address-Size Attributes for Stack Accesses". So, using addrsize to choose the push or pop variant is incorrect in every case.

This is also true. If we knew what the original system addrsize was, then we could use the correct stackaddrsize as well.

**Comment by nneonneo** [comment]
Yes, what I mean is that addrsize should not be used at all. In that case, it doesn't matter whether the instruction knows the original value of addrsize or not. Only stackaddrsize matters, and that's an entirely separate bit from addrsize.

opsize and addrsize are affected by the default bitness of the CPU as well as the D bit of the code segment. stackaddrsize is affected by the B bit of the stack segment. It's completely possible to mix different addrsize and stackaddrsize, for example, when executing 16-bit code with a "big" stack segment (B bit set). 

I see two possible ways to implement this properly. Option one is to add a new persistent context bit to track the stack address size, which is initialized in the pspec appropriately (same as addrsize), and which can therefore be modified by the analyst. Option two is to add a new temporary context bit which is copied from addrsize at the start of each instruction, but is not affected by 0x67. The latter doesn't permit the stack address size to be changed independently of address size, but is backwards compatible (I'm assuming that new pspec context_set commands aren't applied to existing databases).

**Comment by Wall-AF** [comment]
> I see two possible ways to implement this properly. Option one is to add a new persistent context bit to track the stack address size, which is initialized in the pspec appropriately (same as addrsize), and which can therefore be modified by the analyst. Option two is to add a new temporary context bit which is copied from addrsize at the start of each instruction, but is not affected by 0x67. The latter doesn't permit the stack address size to be changed independently of address size, but is backwards compatible (I'm assuming that new pspec context_set commands aren't applied to existing databases).

Please don't implement option 2!

The 3rd option is to have a per application "context bit to track the stack address size".
